torchrl/gradient_collector.py

```diff
@@ -0,0 +1,168 @@
+        updates_per_batch: int = 1,
+        device: torch.device = "cpu",
+    ):
+
+        self.device = device
+        self.updates_per_batch = updates_per_batch
+
+        # Get Actor Critic instance
+        # Could be instantiated passing class + params if necessary
+        self.policy = policy  # TODO: do I really need the policy if we can get the params from objective ?
+        self.policy_params = TensorDict(dict(self.policy.named_parameters()), [])
+        self.critic = critic  # TODO: do I really need the critic if we can get the params from objective ?
+        self.critic_params = TensorDict(dict(self.critic.named_parameters()), [])
+
+        # Get loss instance
+        # Could be instantiated passing class + params if necessary
+        self.objective = objective
+        self.objective_params = TensorDict(dict(self.objective.named_parameters()), [])
+
+        # Get collector instance.
+        # Could be instantiated passing class + params if necessary.
+        self.collector = collector
+
+        # Get storage instance.
+        # Could be instantiated passing class + params if necessary
+        self.replay_buffer = replay_buffer
+
+        # Get storage instance.
+        # Could be instantiated passing class + params if necessary
+        self.optimizer = optimizer
+
+        # Check if policy parameters are being tracked by optimizer
+        for name, param in policy.named_parameters():
+            if optimizer.state.get(name) is None:
+                warnings.warn(f"{name} is NOT being tracked by the optimizer")
+
+        # Check if critic parameters are being tracked by optimizer
+        if self.critic is not None:
+            for name, param in critic.named_parameters():
+                if optimizer.state.get(name) is None:
+                    warnings.warn(f"{name} is NOT being tracked by the optimizer")
+
+    def __iter__(self) -> Iterator[TensorDictBase]:
+        return self.iterator()
+
+    def next(self):
+        try:
+            if self._iterator is None:
+                self._iterator = iter(self)
+            out = next(self._iterator)
+            # if any, we don't want the device ref to be passed in distributed settings
+            out.clear_device_()
+            return out
+        except StopIteration:
+            return None
+
+    def update_policy_weights_(
+            self,
+            policy_weights: Optional[TensorDictBase] = None,
+            critic_weights: Optional[TensorDictBase] = None,
+            objective_weights: Optional[TensorDictBase] = None,
+    ) -> None:
+
+        # TODO: is this the right way to do it ?
+        # Update policy and critic
+        self.objective_params.apply(lambda x: x.data).update_(objective_weights)
+        self.policy_params.apply(lambda x: x.data).update_(policy_weights)
+        self.critic_params.apply(lambda x: x.data).update_(critic_weights)
+
+        for name, param in self.objective.named_parameters():
```

Comment:
in theory your objective params and buffers should match in identity those of the policy and critic
One way to make sure that it's the case would be to register only the objective params and buffers, and during this (in `__init__`) check that they match in identity those of the policy and critic, smth like
```python
if set(policy.parameters()).union(set(critic.parameters())) != objective.parameters():
    raise ValueError
```